typing/typecore.ml

```diff
@@ -975,11 +980,14 @@ let solve_Ppat_record_field ~refine loc penv label label_lid record_ty =
   end
 
 let solve_Ppat_array ~refine loc env expected_ty =
-  let ty_elt = newgenvar() in
   let expected_ty = generic_instance expected_ty in
-  unify_pat_types_refine ~refine
-    loc env (Predef.type_array ty_elt) expected_ty;
-  ty_elt
+  if is_principal expected_ty && is_floatarray_type !!env expected_ty then
+    instance Predef.type_float
```

Comment:
Sorry to nitpick, but this bit is *still* repeated. :)

Maybe something like

```
(* Returns [Some ty_elt] if the expected type can be used for type-directed disambiguation
   of the array literal. *)
let disambiguate_array_literal env expected_ty =
  if is_principal expected_ty then begin
    if is_floatarray_type env expected_ty then Some (instance Predef.type_float) else
    None
  end else None
```

and then this can be used here and for expressions. The structure is designed to be easily extensible with new types that might participate in this.